man/man1/tail.1
.\" @(#)tail.1 1.1 92/07/30 SMI; from S5R2 6.2 83/09/02
.TH TAIL 1 "14 August 1989"
.SH NAME
tail \- display the last part of a file
.SH SYNOPSIS
.B tail
\fB+\|\fR|\fB\|\-\c
.I number
[
.B lbc
] [
.B f
] [
.I filename
]
.br
.B tail
\fB+\|\fR|\fB\|\-\c
.I number
[
.B l
] [
.B rf
] [
.I filename
]
.IX  "tail command"  ""  "\fLtail\fP \(em display last part of file"
.IX  display "last part of file \(em \fLtail\fP"
.IX  file  "display last part of"  ""  "display last part of \(em \fLtail\fP"
.IX  "text processing utilities"  tail  ""  "\fLtail\fP \(em display last part of file"
.SH DESCRIPTION
.B tail
copies
.I filename
to the standard output beginning at a
designated place.  If no file is named, the standard input is used.
.SH OPTIONS
.LP
Options are all jammed together, not specified separately with their
own
.RB ` \- '
signs.
.TP
.BI \+ number
Begin copying at distance
.I number
from the beginning of the file.
.I number
is counted in units of lines, blocks or
characters,
according to the appended option
.BR l ,
\fBb\fP, or
.BR c .
When no units are specified, counting is by lines.
If
.I number
is not specified, the value 10 is used.
.TP
.BI \- number
Begin copying at distance
.I number
from the end of the file.
.I number
is counted in units of lines, blocks or
characters,
according to the appended option
.BR l ,
\fBb\fP, or
.BR c .
When no units are specified, counting is by lines.
If
.I number
is not specified, the value 10 is used.
.TP
.B l
.I number
is counted in units of lines.
.TP
.B b
.I number
is counted in units of blocks.
.TP
.B c
.I number
is counted in units of characters.
.TP
.B r
Copy lines from the end of the file in
reverse order.  The default for option
.B r
is to print the entire file in reverse order.
.I number
is the count of lines from the end of the file regardless of sign.
Option
.B r
may not be used with
.B b
or
.BR c .
.TP
.B f
If the input file is not a pipe,
do not terminate after the line of the input
file has been copied, but enter an endless loop,
sleeping for a second and then attempting to read and copy
further records from the input file.
This option may be used to monitor the growth of a file that is
being written by some other process.
For example, the command:
.RS
.IP
.B tail \|\-f \|fred
.RE
.IP
will print the last ten lines of the file
.BR fred ,
followed by any lines that are appended to
.B fred
between the time
.B tail
is initiated and killed.
As another example, the command:
.RS
.IP
.B tail \|\-15cf \|fred
.RE
.IP
will print the last 15 characters of the file
.BR fred ,
followed by any lines that are appended to
.B fred
between the time
.B tail
is initiated and killed.
.SH "SEE ALSO"
.BR dd (1)
.SH BUGS
Data for a tail relative to the end of the file is stored in a buffer,
and thus is limited in size.
.LP
Various kinds of anomalous behavior may happen with character special
files.


ucb/tail.c
#ifndef lint
static	char sccsid[] = "@(#)tail.c 1.1 92/07/30 SMI"; /* from UCB 5.2 1/10/86 */
#endif

/*
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1980 Regents of the University of California.\n\
 All rights reserved.\n";
#endif not lint

/* tail command 
 *
 *	tail where [filename]
 *	where is [+/-]n[type]
 *	- means n lines before end
 *	+ means nth line from beginning
 *	type 'b' means tail n blocks, not lines
 *	type 'c' means tail n characters
 *	Type 'r' means in lines in reverse order from end
 *	 (for -r, default is entire buffer )
 *	option 'f' means loop endlessly trying to read more
 *		characters after the end of file, on the assumption
 *		that the file is growing
*/

#include	<stdio.h>
#include	<ctype.h>
#include	<sys/types.h>
#include	<sys/stat.h>
#include	<sys/file.h>
#include	<errno.h>

#define LBIN 32769
#undef	BUFSIZ
#define	BUFSIZ	8192
struct	stat	statb;
int	follow;
int	piped;
char bin[LBIN];
int errno;

main(argc,argv)
char **argv;
{
	long n,di;
	register i,j,k;
	char	*arg;
	int partial, bylines, bkwds, fromend, lastnl;
	char *p;

	arg = argv[1];
	if (argc<=1 || *arg!='-' && *arg!='+') {
		arg = "-10l";
		argc++;
		argv--;
	}
	fromend = *arg=='-';
	arg++;
	if (isdigit(*arg)) {
		n = 0;
		while(isdigit(*arg))
			n = n*10 + *arg++ - '0';
	} else
		n = -1;
	if (!fromend && n>0)
		n--;
	if (argc>2) {
		(void)close(0);
		if (open(argv[2],0)!=0) {
			fprintf(stderr, "tail: ");
			perror(argv[2]);
			exit(1);
		}
	}
	(void)lseek(0,(off_t)0,L_INCR);
	piped = errno==ESPIPE;
	bylines = -1; bkwds = 0;
	while(*arg)
	switch(*arg++) {

	case 'b':
		if (n == -1) n = 1;
		n <<= 9;
		if (bylines!=-1 || bkwds==1) goto errcom;
		bylines=0;
		break;
	case 'c':
		if (bylines!=-1 || bkwds==1) goto errcom;
		bylines=0;
		break;
	case 'f':
		follow = 1;
		break;
	case 'r':
		if (n==-1) n = LBIN;
		else if (!fromend) n++;
		if (bylines==0) goto errcom;
		bkwds = 1; fromend = 1; bylines = 1;
		break;
	case 'l':
		if (bylines!=-1 && bylines!=1) goto errcom;
		bylines = 1;
		break;
	default:
		goto errcom;
	}
	if (n==-1) n = 10;
	if (bylines==-1) bylines = 1;
	if (bkwds) follow=0;
	if (fromend)
		goto keep;

			/*seek from beginning */

	if (bylines) {
		j = 0;
		while(n-->0) {
			do {
				if (j--<=0) {
					p = bin;
					j = read(0,p,BUFSIZ);
					if (j--<=0)
						fexit();
				}
			} while(*p++ != '\n');
		}
		(void)write(1,p,j);
	} else  if (n>0) {
		if (!piped)
			(void)fstat(0,&statb);
		if (piped || (statb.st_mode&S_IFMT)==S_IFCHR)
			while(n>0) {
				i = n>BUFSIZ?BUFSIZ:n;
				i = read(0,bin,i);
				if (i<=0)
					fexit();
				n -= i;
			}
		else
			(void)lseek(0,(off_t)n,L_SET);
	}
copy:
	while((i=read(0,bin,BUFSIZ))>0)
		(void)write(1,bin,i);
	fexit();

			/*seek from end*/

keep:
	if (n <= 0)
		fexit();
	if (!piped) {
		(void)fstat(0,&statb);
		/* If by lines, back up 1 buffer: else back up as needed */
		di = bylines?LBIN-1:n;
		if (statb.st_size > di)
			(void)lseek(0,(off_t)-di,L_XTND);
		if (!bylines)
			goto copy;
	}
	partial = 1;
	for(;;) {
		i = 0;
		do {
			j = read(0,&bin[i],LBIN-i);
			if (j<=0)
				goto brka;
			i += j;
		} while(i<LBIN);
		partial = 0;
	}
brka:
	if (!bylines) {
		k =
		    n<=i ? i-n:
		    partial ? 0:
		    n>=LBIN ? i+1:
		    i-n+LBIN;
		k--;
	} else {
		if (bkwds && bin[i==0?LBIN-1:i-1]!='\n'){	/* force trailing newline */
			bin[i]='\n';
			if (++i>=LBIN) {i = 0; partial = 0;}
		}
		k = i;
		j = 0;
		do {
			lastnl = k;
			do {
				if (--k<0) {
					if (partial) {
						if (bkwds) 
						    (void)write(1,bin,lastnl+1);
						goto brkb;
					}
					k = LBIN -1;
				}
			} while(bin[k]!='\n' && k!=i);
			if (bkwds && j>0){
				if (k<lastnl) (void)write(1,&bin[k+1],lastnl-k);
				else {
					(void)write(1,&bin[k+1],LBIN-k-1);
					(void)write(1,bin,lastnl+1);
				}
			}
		} while(j++<n && k!=i);
brkb:
		if (bkwds) exit(0);
		if (k==i) do {
			if (++k>=LBIN)
				k = 0;
		} while(bin[k]!='\n' && k!=i);
	}
	if (k<i)
		(void)write(1,&bin[k+1],i-k-1);
	else {
		(void)write(1,&bin[k+1],LBIN-k-1);
		(void)write(1,bin,i);
	}
	fexit();
errcom:
	fprintf(stderr, "usage: tail [+/-[n][lbc][f]] [filename]\n");
	fprintf(stderr, "       tail [+/-[n][l][rf]] [filename]\n");
	exit(2);
	/* NOTREACHED */
}

fexit()
{	register int n;
	if (!follow || piped) exit(0);
	for (;;)
	{	sleep(1);
		while ((n = read (0, bin, BUFSIZ)) > 0)
			(void)write (1, bin, n);
	}
}
